cpp/src/parquet/file_reader.cc

```diff
@@ -264,23 +264,92 @@ class SerializedFile : public ParquetFileReader::Contents {
       }
     }
     PARQUET_THROW_NOT_OK(cached_source_->Cache(ranges));
-    return cached_source_->Wait();
   }
 
+  ::arrow::Future<> WhenBuffered(const std::vector<int>& row_groups,
+                                 const std::vector<int>& column_indices) const {
+    if (!cached_source_) {
+      return ::arrow::Status::Invalid("Must call PreBuffer before WhenBuffered");
+    }
+    std::vector<::arrow::io::ReadRange> ranges;
+    for (int row : row_groups) {
+      for (int col : column_indices) {
+        ranges.push_back(
+            ComputeColumnChunkRange(file_metadata_.get(), source_size_, row, col));
+      }
+    }
+    return cached_source_->WaitFor(ranges);
+  }
+
+  // Metadata/footer parsing. Divided up to separate sync/async paths, and to use
+  // exceptions for error handling (with the async path converting to Future/Status).
+
   void ParseMetaData() {
```

Comment:
Can we express this in terms of `ParseMetadataAsync`? I would rather not have duplicate code paths for this.

Reply:
Originally I did do that but found I was having issues with the encryption test flaking on Windows. I couldn't see why this change affected that (since we wait for the future to complete) but refactoring it like this fixed it. Also, this means we don't need a custom StatusDetail to preserve the original exception class thrown.

Reply:
Hmm, ok, I see. Is the async path for encryption tested somewhere?

Reply:
I've updated the current encryption tests to test both Open and OpenAsync (though let's hope they don't make the test flaky again).